CARDS-1417: Print-friendly styling of forms

marta- opened the pull request:
To test: create a form, enter some data, save and view, and use the newly introduced "Print" menu entry to launch the browser's  print functionality.

marta- committed:
CARDS-1417: Print-friendly styling of forms
* Hide menus
* Improve alignment
* Set page breaking rules

marta- committed:
CARDS-1417: Print-friendly styling of forms
Added a Print menu entry in view mode, which launches the browser's print functionality

marta- committed:
[misc] Only save before txt export if we're in edit mode

marta- committed:
[misc] Provide the "Change subject" menu entry for forms only in edit mode. In view mode it may be mistaken for a navigation feature.

marta- committed:
CARDS-1417: Print-friendly styling of forms
Bugfix: form header appears collapsed and in the wrong location when the form is scrolled up before clicking Print

sashaandjic commented:
- IDs shown instead of vocabulary terms in print and in export
- Problem with pedigree in the print (see the attachment)
- If possible it would be nice to have page numbers and header shown on each page of the print document 
[ABC-123 Export.txt](https://github.com/data-team-uhn/cards/files/7343551/ABC-123.Export.txt)
[ABC-123_ Patient information _ LFS Data Core.pdf](https://github.com/data-team-uhn/cards/files/7343552/ABC-123_.Patient.information._.LFS.Data.Core.pdf)

marta- committed:
CARDS-1417: Print-friendly styling of forms
Quick fix for pedigree thumbnail overlapping the next question

marta- commented:
> LFS mode:
> 
> * IDs shown instead of vocabulary terms in print and in export

Looks like a regression. It also happens in view mode so it's not specific to export or print. Should be reported separately.

> * Problem with pedigree in the print (see the attached PDF)

Quick fix in 469392a3. Eventually we may want to print pedigrees and other images one per page at the end of the form, but this will do for now.

> * If possible it would be nice to have page numbers and header shown on each page of the print document

We are using the browser native print function. While having custom headers is theoretically possible, it is not a priority at this time. Page numbers can be enabled from the print settings once the print dialog is displayed.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -354,15 +354,24 @@ function Form (props) {
-                        Change subject
-                      </Button>
-                    </ListItem>
+                    { isEdit &&
```
Why not have the change subject in view mode?

marta- replied:
9b6d651 : In view mode it can easily be mistaken for a navigation feature.

sashaandjic reviewed (approved):
tested
[ABC-!@#-(012)_ Patient information _ LFS Data Core.pdf](https://github.com/data-team-uhn/cards/files/7347214/ABC-.%40.-.012._.Patient.information._.LFS.Data.Core.pdf)
[export as text.txt](https://github.com/data-team-uhn/cards/files/7347215/export.as.text.txt)

sdumitriu reviewed (commented):
- In FF, I can get a completely blank page at the end.
- In Chrome, I can get weird spacing, see the attached files.
- I think we do need some branded header or footer, it looks really bland and there's no indication where a file is coming from.
- There is only one link in the whole document, for the subject. Should we remove the underline and color from it? Should we add a link to the form name as well, `Patient Information` leads back to the form.

[print-chrome.pdf](https://github.com/data-team-uhn/cards/files/7347425/print-chrome.pdf)
[print-firefox.pdf](https://github.com/data-team-uhn/cards/files/7347426/print-firefox.pdf)

sashaandjic commented:
> * I think we do need some branded header or footer, it looks really bland and there's no indication where a file is coming from.
> * Should we add a link to the form name as well, `Patient Information` leads back to the form.
 
These 2 suggestions would be nice to have.

marta- committed:
CARDS-1417: Print-friendly styling of forms
Bug fix: prevent the form menu dropdown from being accidentally displayed in print preview

marta- commented:
> > Bug: the dropdown menu (containing Print, Export as text, Delete) remains visible in the printed version on Chrome/Windows.
> 
> Can you provide more details or a screenshot? [ABC-!@#-(012)_ Patient information _ LFS Data Core.pdf](https://github.com/data-team-uhn/cards/files/7391677/ABC-.%40.-.012._.Patient.information._.LFS.Data.Core.pdf)

It's not easy to replicate, it only happened once actually. Sometimes when clicking More actions > Print, the print preview is generated before the dropdown menu gets hidden. Commit 3d023db adds a CSS rule that hides it for `@media print`.

marta- committed:
CARDS-1417: Print-friendly styling of forms
Added a text-only breadcrumb for print mode

marta- committed:
CARDS-1417: Print-friendly styling of forms
Better look for header/footer sections in print mode

marta- commented:
Summary of latest changes:
* Rebased on `dev`
* Replaced the breadcrumbs with links (Patient [123](#) / Visit [456](#)) with plain text breadcrumbs in print mode
* Improved the look of header/footer sections in print mode

sashaandjic commented:
Probably out of scope: Can we hide this when in view mode?
![image](https://user-images.githubusercontent.com/23243244/150408159-14c71b19-05fe-43bf-84d4-a5b10a9a8d2d.png)
![image](https://user-images.githubusercontent.com/23243244/150408192-a86bcc12-3b48-4ef1-999e-63c7a9b106c9.png)

sashaandjic commented:
Another out of scope: Status may look better if it is right next to the form's title. Or perhaps next to the description?
![image](https://user-images.githubusercontent.com/23243244/150408266-1eee13c3-6546-4920-8817-5629145c3eaa.png)
![image](https://user-images.githubusercontent.com/23243244/150408444-07d9b663-dc11-4edf-b54d-5c736c24f07e.png)

sashaandjic commented:
The sentence at the bottom: "If your symptoms result..." is not shown when using `Print` option.
![image](https://user-images.githubusercontent.com/23243244/150409216-f3d941fb-7daa-4247-9bba-329fa0399601.png)
[1234567 _ 11_ Patient Health Questionnaire-9 _ ePRO.pdf](https://github.com/data-team-uhn/cards/files/7908001/1234567._.11_.Patient.Health.Questionnaire-9._.ePRO.pdf)

marta- commented:
> Probably out of scope: Can we hide this when in view mode? !

Added to https://phenotips.atlassian.net/browse/CARDS-1355 todos.

marta- commented:
> Another out of scope: Status may look better if it is right next to the form's title. Or perhaps next to the description? In red fonts? 

This is https://phenotips.atlassian.net/browse/CARDS-1541.

marta- commented:
> The sentence at the bottom: "If your symptoms result..." is not shown when using `Print` option for all but SC form.

That's a "summary" section that should be excluded from text export. Please report as a separate task.

sdumitriu commented:
Some really ugly things (near the end):
[firefox.pdf](https://github.com/data-team-uhn/cards/files/7908975/firefox.pdf)
[chrome.pdf](https://github.com/data-team-uhn/cards/files/7908981/chrome.pdf)

sashaandjic commented:
We should change in the pdfs default name "ePro" to "DATA-PRO"
![image](https://user-images.githubusercontent.com/23243244/150445478-fe8411af-c8c2-494f-a50d-c9eda5b0ef74.png)

marta- commented:
> We should change in the pdfs default name "ePro" to "DATA-PRO" ![image](https://user-images.githubusercontent.com/23243244/150445478-fe8411af-c8c2-494f-a50d-c9eda5b0ef74.png)

This branch doesn't include 1563, which changed the app name. It will look fine once debases on the latest dev

sashaandjic commented:
> Some really ugly things (near the end):

On Windows, in LFS mode, it does not look  that bad. There is a little bit of overlap in one spot on both. See the attached pdfs. In Proms mode I found no issues.
[mozilaWin.pdf](https://github.com/data-team-uhn/cards/files/7909542/mozilaWin.pdf) - see page 2 
![image](https://user-images.githubusercontent.com/23243244/150447692-57983c74-c764-4360-83f0-c59ce82934fe.png)

[chromeWin.pdf](https://github.com/data-team-uhn/cards/files/7909544/chromeWin.pdf) - see page 6 
![image](https://user-images.githubusercontent.com/23243244/150447716-3527cf72-f150-4f98-8717-b2bc6bdd38a2.png)

sashaandjic commented:
Can we keep together the field label and its content, so they do not get broken into different pages?

marta- commented:
Closing in favor of #866 .
